File: ruby/9a.rb
#!/usr/bin/ruby

# Ignore white space and split into characters
input = File.open('9.in').read.gsub(/\s/, '').chars

out = []

# While there is input left..
while input.any?
  case input.first
  when '('
    # If it's an escape character, consume the escape seqence
    i = input.take_while{|x| x != ')'}.join
    # Extract the RLE length and times
    x, y = i.scan(/\d+/).map(&:to_i)
    # Remove escape from imput
    input = input.drop(i.size + 1)
    # Add to the output
    out <<= input.take(x) * y
    # Remove RLE from the imput
    input = input.drop(x)
  else
    # Find the non RLE content
    i = input.take_while{|x| x != '('}
    # Add to the output
    out <<= i
    # Remove from the input
    input = input.drop(i.size)
  end
end

puts out.flatten.size

File: python/9a.py
#!/usr/bin/python

def decompress(data):
    """
    Iterate through an iterable type e.g. string or character
    array looking for RLE control sequences and expanding the
    next N characters M times.  Return the length.
    """
    length = 0
    iterator = iter(data)
    for char in iterator:
        # Control sequence
        if char == '(':
            # Collect the control fields
            control = ""
            while True:
                peek = iterator.next()
                if peek == ')':
                    break
                control += peek
            take, times = [int(x) for x in control.split('x')]
            # Extract 'take' characters
            [iterator.next() for x in range(0, take)]
            # Accumulate 'take' characters 'times' times
            length += take * times
        else:
            length += 1
    return length


def main():
    # Load the input
    data = open('9.in').read().strip()
    # Answer is the length of the output
    print decompress(data)


if __name__ == '__main__':
    main()

# vi: ts=4 et:
